llvm/lib/Transforms/IPO/FunctionAttrs.cpp

```diff
@@ -2322,8 +2352,29 @@ PreservedAnalyses PostOrderFunctionAttrsPass::run(LazyCallGraph::SCC &C,
     Functions.push_back(&N.getFunction());
   }
 
+  bool AnyFunctionsAddressIsTaken = false;
+  // Check if any function in the whole program has its address taken.
+  // We use this information when inferring norecurse attribute: If there is
+  // no function whose address is taken, we conclude that any external function
+  // cannot callback into any user function.
```

Comment:
Being in LTO post-link does *not* imply that you have whole program visibility. To give a simple example, you could be linking a dylib. LTO post-link is still subject to an export list. Only the symbols which have been internalized are internal.

Reply:
Updated to assume address is taken when there exists any function which doesn't have private/internal linkage.